ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/BatchedSubmissionValidator.scala

```diff
@@ -0,0 +1,439 @@
+          .readState(inputKeys)
+          .map { values =>
+            (correlatedSubmission, inputKeys.zip(values).toMap)
+          }
+      )
+    }
+  }
+
+  private def validateSubmission(
+      participantId: ParticipantId,
+      recordTime: Timestamp,
+      correlatedSubmission: CorrelatedSubmission,
+      inputState: DamlInputState)(
+      implicit executionContext: ExecutionContext): Future[ValidatedSubmission] =
+    withSubmissionLoggingContext(correlatedSubmission) { implicit logCtx =>
+      Timed.timedAndTrackedFuture(
+        Metrics.validate,
+        Metrics.validateRunning,
+        Future {
+          val logEntryAndState = committer.processSubmission(
+            correlatedSubmission.logEntryId,
+            recordTime,
+            LedgerReader.DefaultConfiguration,
+            correlatedSubmission.submission,
+            participantId,
+            inputState
+          )
+          (correlatedSubmission, inputState, logEntryAndState)
+        }
+      )
+    }
+
+  private def detectConflictsAndRecover(
+      correlatedSubmission: CorrelatedSubmission,
+      inputState: DamlInputState,
+      logEntryAndState: LogEntryAndState,
+      invalidatedKeys: mutable.Set[DamlStateKey])
+    : scala.collection.immutable.Iterable[ValidatedSubmission] = {
+    val (logEntry, outputState) = logEntryAndState
+    withSubmissionLoggingContext(correlatedSubmission) { implicit logCtx =>
+      Timed.value(
+        Metrics.detectConflicts, {
+          conflictDetection
+            .detectConflictsAndRecover(
+              invalidatedKeys,
+              inputState,
+              logEntry,
+              outputState
+            )
+            .map {
+              case (newInvalidatedKeys, (newLogEntry, newState)) =>
+                invalidatedKeys ++= newInvalidatedKeys
+                (correlatedSubmission, inputState, (newLogEntry, newState)) :: Nil
+            }
+            .getOrElse {
+              logger.info(
+                s"Submission ${correlatedSubmission.correlationId} dropped as it conflicted and recovery was not possible")
+              Nil
+            }
+        }
+      )
+    }
+  }
+
+  private def commitResult(
+      participantId: ParticipantId,
+      correlatedSubmission: CorrelatedSubmission,
+      inputState: DamlInputState,
+      logEntryAndState: LogEntryAndState,
+      commitStrategy: CommitStrategy[CommitResult])(
+      implicit executionContext: ExecutionContext): Future[Unit] = {
+    val (logEntry, outputState) = logEntryAndState
+    withSubmissionLoggingContext(correlatedSubmission) { implicit logCtx =>
+      Timed.timedAndTrackedFuture(
+        Metrics.commit,
+        Metrics.commitRunning,
+        commitStrategy
+          .commit(
+            participantId,
+            correlatedSubmission.correlationId,
+            correlatedSubmission.logEntryId,
+            logEntry,
+            inputState,
+            outputState)
+          .map(_ => ())
```

Comment:
This gets a `CommitResult`, but that is then discarded. If we were to propagate this outwards, we can use this in the `InMemoryLedgerStateOperations` instead of getting the `log.size` again.

Reply:
Yes, that's intentional because this way we don't have to deal with multiple `CommitResult`s. I'm happy to add this in a separate PR, though, e.g., by taking the max of the `CommitResult`s and returning that instead of `Unit`. Does that work for you?

Reply:
Sounds good. I expect writing an API for some kind of reduce function will be quite a considerable change.

Reply:
Agreed. To avoid that I'd first go with adding some default reduce function (e.g., max as the log offsets should be monotonically increasing anyway) and in case something else is needed one can handle that as part of the `CommitStrategy` and/or `LedgerWriter` implementation. For production ledgers the signaling mechanism is not necessarily tied to the validation logic as in case of the in-memory or SQL ledgers (e.g., the participant will be notified of new blocks after they have been committed which drives the dispatcher instead of the validation logic in a post-commit step).

Reply:
You may not be able to do that; you don't know anything about `CommitStrategy`. But it doesn't matter for now.

Reply:
Yeah, that's right, however, I was describing from the consumer's side, i.e., as you can pass in a separate `CommitStrategy` instance for each `validateAndCommit` call you can simply collect the log results and do something with them once the validation is finished.